<b>Português:</b>  Uso do Pandas para investigar um conjunto de dados, no período de 2008 à 2018, sobre a economia de combustível em diferentes modelos de carros ( all_alpha_08.csv e all_alpha_18.csv).<br> 

Características dos conjuntos de dados avaliadas:<br> 

-número de amostras em cada conjunto<br> 
-número de colunas em cada conjunto<br> 
-linhas duplicadas em cada conjunto<br> 
-tipos de dados das colunas<br> 
-atributos com valor ausente<br> 
-número de valores únicos não nulos para atributos em cada conjunto.<br> 

<b>English:</b>  Using Pandas to investigate a dataset, from 2008 to 2018, on fuel economy in different car models (all_alpha_08.csv and all_alpha_18.csv).<br> 

Characteristics of the evaluated data sets:<br> 

-number of samples in each set<br> 
-number of columns in each set<br> 
-duplicate rows in each set<br> 
-data types of columns<br> 
-attributes with missing values<br> 
-number of unique non-zero values for attributes in each set.<br> 

In [31]:
import pandas as pd
df_08 = pd.read_csv("all-alpha-08.csv")
df_18 = pd.read_csv("all-alpha-18.csv")

In [32]:
print("df_08")
display(df_08.head())
print("df_18")
display(df_18.head())

df_08


Model  Displ      Cyl    Trans Drive      Fuel Sales Area Stnd  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline         CA   U2   
1  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline         FA   B5   
2  ACURA RDX    2.3  (4 cyl)  Auto-S5   4WD  Gasoline         CA   U2   
3  ACURA RDX    2.3  (4 cyl)  Auto-S5   4WD  Gasoline         FA   B5   
4   ACURA RL    3.5  (6 cyl)  Auto-S5   4WD  Gasoline         CA   U2   

   Underhood ID    Veh Class Air Pollution Score FE Calc Appr City MPG  \
0  8HNXT03.7PKR          SUV                   7          Drv       15   
1  8HNXT03.7PKR          SUV                   6          Drv       15   
2  8HNXT02.3DKR          SUV                   7          Drv       17   
3  8HNXT02.3DKR          SUV                   6          Drv       17   
4  8HNXV03.5HKR  midsize car                   7          Drv       16   

  Hwy MPG Cmb MPG  Unadj Cmb MPG Greenhouse Gas Score SmartWay  
0      20      17        22.0527                    4       no  
1      20      17        22.0527                    4       no  
2      22      19        24.1745                    5       no  
3      22      19        24.1745                    5       no  
4      24      19        24.5629                    5       no

df_18


Model  Displ  Cyl       Trans Drive      Fuel Cert Region       Stnd  \
0  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline          FA     T3B125   
1  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline          CA         U2   
2  ACURA RDX    3.5  6.0  SemiAuto-6   4WD  Gasoline          FA     T3B125   
3  ACURA RDX    3.5  6.0  SemiAuto-6   4WD  Gasoline          CA         U2   
4  ACURA TLX    2.4  4.0       AMS-8   2WD  Gasoline          CA  L3ULEV125   

             Stnd Description  Underhood ID  Veh Class  Air Pollution Score  \
0      Federal Tier 3 Bin 125  JHNXT03.5GV3  small SUV                    3   
1      California LEV-II ULEV  JHNXT03.5GV3  small SUV                    3   
2      Federal Tier 3 Bin 125  JHNXT03.5GV3  small SUV                    3   
3      California LEV-II ULEV  JHNXT03.5GV3  small SUV                    3   
4  California LEV-III ULEV125  JHNXV02.4WH3  small car                    3   

  City MPG Hwy MPG Cmb MPG  Greenhouse Gas Score SmartWay Comb CO2  
0       20      28      23                     5       No      386  
1       20      28      23                     5       No      386  
2       19      27      22                     4       No      402  
3       19      27      22                     4       No      402  
4       23      33      27                     6       No      330

In [33]:
df_08.shape

(2404, 18)

In [34]:
df_18.shape

(1611, 18)

In [35]:
print("Numero de amostras - 2008: {0}".format(df_08.shape[0])) #number of samples on 2008 data set

Numero de amostras - 2008: 2404


In [36]:
print("Numero de amostras - 2018: {0}".format(df_18.shape[0])) #number of samples on 2018 data set

Numero de amostras - 2018: 1611


In [37]:
print("Numero de colunas - 2008: {0}".format(df_08.shape[1])) #number of columns on 2008 data set

Numero de colunas - 2008: 18


In [38]:
print("Numero de colunas - 2018: {0}".format(df_18.shape[1])) #number of samples on 2018 data set

Numero de colunas - 2018: 18


In [39]:
df_08.duplicated().sum()

25

In [40]:
print("Numero de linhas duplicadas - 2008: {0}".format(df_08.duplicated().sum())) #number of duplicated lines on 2008 data set

Numero de linhas duplicadas - 2008: 25


In [41]:
df_18.duplicated().sum()

0

In [42]:
print("Numero de linhas duplicadas - 2018: {0}".format(df_18.duplicated().sum())) #number of duplicated lines on 2018 data set

Numero de linhas duplicadas - 2018: 0


In [43]:
df_08.dtypes 

Model                    object
Displ                   float64
Cyl                      object
Trans                    object
Drive                    object
Fuel                     object
Sales Area               object
Stnd                     object
Underhood ID             object
Veh Class                object
Air Pollution Score      object
FE Calc Appr             object
City MPG                 object
Hwy MPG                  object
Cmb MPG                  object
Unadj Cmb MPG           float64
Greenhouse Gas Score     object
SmartWay                 object
dtype: object

In [44]:
df_18.dtypes 

Model                    object
Displ                   float64
Cyl                     float64
Trans                    object
Drive                    object
Fuel                     object
Cert Region              object
Stnd                     object
Stnd Description         object
Underhood ID             object
Veh Class                object
Air Pollution Score       int64
City MPG                 object
Hwy MPG                  object
Cmb MPG                  object
Greenhouse Gas Score      int64
SmartWay                 object
Comb CO2                 object
dtype: object

In [45]:
def check_types(df): #data types of columns on data set 2008
    for column in df.columns:
        print("coluna: {0}, tipo: {1}".format(column, type(df[column][0])))
check_types(df_08)

coluna: Model, tipo: <class 'str'>
coluna: Displ, tipo: <class 'numpy.float64'>
coluna: Cyl, tipo: <class 'str'>
coluna: Trans, tipo: <class 'str'>
coluna: Drive, tipo: <class 'str'>
coluna: Fuel, tipo: <class 'str'>
coluna: Sales Area, tipo: <class 'str'>
coluna: Stnd, tipo: <class 'str'>
coluna: Underhood ID, tipo: <class 'str'>
coluna: Veh Class, tipo: <class 'str'>
coluna: Air Pollution Score, tipo: <class 'str'>
coluna: FE Calc Appr, tipo: <class 'str'>
coluna: City MPG, tipo: <class 'str'>
coluna: Hwy MPG, tipo: <class 'str'>
coluna: Cmb MPG, tipo: <class 'str'>
coluna: Unadj Cmb MPG, tipo: <class 'numpy.float64'>
coluna: Greenhouse Gas Score, tipo: <class 'str'>
coluna: SmartWay, tipo: <class 'str'>


In [46]:
check_types(df_18) #data types of columns on data set 2018

coluna: Model, tipo: <class 'str'>
coluna: Displ, tipo: <class 'numpy.float64'>
coluna: Cyl, tipo: <class 'numpy.float64'>
coluna: Trans, tipo: <class 'str'>
coluna: Drive, tipo: <class 'str'>
coluna: Fuel, tipo: <class 'str'>
coluna: Cert Region, tipo: <class 'str'>
coluna: Stnd, tipo: <class 'str'>
coluna: Stnd Description, tipo: <class 'str'>
coluna: Underhood ID, tipo: <class 'str'>
coluna: Veh Class, tipo: <class 'str'>
coluna: Air Pollution Score, tipo: <class 'numpy.int64'>
coluna: City MPG, tipo: <class 'str'>
coluna: Hwy MPG, tipo: <class 'str'>
coluna: Cmb MPG, tipo: <class 'str'>
coluna: Greenhouse Gas Score, tipo: <class 'numpy.int64'>
coluna: SmartWay, tipo: <class 'str'>
coluna: Comb CO2, tipo: <class 'str'>


Abaixo o cálculo da quantidade de linhas sem dados em cada tabela:<br> 
(Obs. para colunas o código seria df.isnull().any())

Below is the calculation of the number of rows without data in each data set:<br> 
(Note: for columns the code would be df.isnull().any())

In [47]:
df_08.isnull().any(axis=1).sum()

199

In [48]:
df_18.isnull().any(axis=1).sum()

2

Número de valores únicos não nulos para atributos em cada conjunto. <br> 
Onde cada um desses tipos de combustível se apresenta?

Number of unique non-zero values for attributes in each set. <br>
Where do each of these fuel types appear?

In [49]:
Fuel_options=df_08['Fuel'].nunique()  # number of unique non-zero values for attributes for 2008 data set
display(Fuel_options)

5

In [50]:
Fuel_options=df_18['Fuel'].nunique() # number of unique non-zero values for attributes for 2018 data set
display(Fuel_options)

5

In [51]:
df_08["Fuel"].unique()

array(['Gasoline', 'ethanol/gas', 'ethanol', 'diesel', 'CNG'],
      dtype=object)

In [52]:
df_18["Fuel"].unique() 

array(['Gasoline', 'Gasoline/Electricity', 'Diesel', 'Ethanol/Gas',
       'Electricity'], dtype=object)

Gasoline: 2008, 2018 <br> 
Etanol/gas: 2008, 2018 <br> 
Gasoline/electricity: 2018 <br> 
GNV: nenhum <br> 
Electricity: 2018 <br> 

número de amostras em cada conjunto: df8=2404. df18=1611
número de colunas em cada conjunto:18
linhas duplicadas em cada conjunto:df8=25, df18=0
tipos de dados das colunas: In[17] e In[18] (células acima)
atributos com valor ausente: df8=199, df18=2
número de valores únicos não nulos para atributos em cada conjunto: 5

Fonte dados: https://www.epa.gov/compliance-and-fuel-economy-data/data-cars-used-testing-fuel-economy </b> 

Data source: https://www.epa.gov/compliance-and-fuel-economy-data/data-cars-used-testing-fuel-economy </b>
